subsys/bluetooth/controller/hci/hci.c

```diff
@@ -5717,27 +5718,33 @@ int hci_iso_handle(struct net_buf *buf, struct net_buf **evt)
 		}
 
 		struct ll_conn_iso_group *cig = cis->group;
+		uint8_t event_offset;
 
 		hdr = &(cis->hdr);
 
-		/* Set target event as the current event. This might cause some
-		 * misalignment between SDU interval and ISO interval in the
-		 * case of a burst from the application or late release. However
-		 * according to the specifications:
-		 * BT Core V5.3 : Vol 6 Low Energy Controller : Part B LL Spec:
-		 * 4.5.13.3 Connected Isochronous Data:
-		 * This burst is associated with the corresponding CIS event but
-		 * the payloads may be transmitted in later events as well.
-		 * If flush timeout is greater than one, use the current event,
-		 * otherwise postpone to the next.
+		/* We must ensure sufficient time for ISO-AL to fragment SDU and
+		 * deliver PDUs to the TX queue. By checking ull_ref_get, we
+		 * know if we are within the subevents of an ISO event. If so,
+		 * we can assume that we have enough time to deliver in the next
+		 * ISO event. If we're not active within the ISO event, we don't
+		 * know if there is enough time to deliver in the next event,
+		 * and for safety we set the target to current event + 2.
 		 *
-		 * TODO: Calculate the best possible target event based on CIS
-		 * reference, FT and event_count.
+		 * For FT>1, we have the opportunity to retransmit in later
+		 * event(s), in which case we have the option to target an
+		 * earlier event (this or next) because being late does not
+		 * instantly flush the payload.
 		 */
-		sdu_frag_tx.target_event = cis->lll.event_count +
-			(cis->lll.tx.flush_timeout > 1 ? 0 : 1);
+		event_offset = ull_ref_get(&cig->ull) ? 1 : 2;
 
-		sdu_frag_tx.grp_ref_point = cig->cig_ref_point;
+		if (cis->lll.tx.flush_timeout > 1) {
+			/* FT>1, target an earlier event */
+			event_offset -= 1;
+		}
+		sdu_frag_tx.target_event = cis->lll.event_count + event_offset;
+		sdu_frag_tx.grp_ref_point = isoal_get_wrapped_time_us(cig->cig_ref_point,
+						(event_offset * cig->iso_interval *
+							ISO_INT_UNIT_US));
```

Comment:
Just thinking out aloud here:
I'm wondering if the source generating the audio is drifting with respect to the CIG reference point over time?
Since the CIG reference point is synchronized with the central it will most likely drift with respect to the periodic events of the source generating the SDUs. This would mean that even if the SDUs arrived in time to be fragmented in to payloads that could be transmitted when streaming started, as the CIG reference point drifts, the SDUs might start arriving later and finally too late for the sub-events in the target event.

Reply:
Any progress or updates?

Reply:
I still see loss of ISO data, I have BabbleSim tests for Multiple Central CIS support in draft PR that fail with change in this PR (it could be the changes in the draft PR too, but the tests pass without the changes in this PR).

The draft PR with the tests rebase to changes in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/56074

Reply:
The purpose of these changes are to ensure that we don't skew the stream of data when fragmenting SDUs. This is done by considering the SDU interval as well as the incoming SDU packet SN and time-stamp.

I debugged a bit on the PR that you mentioned. I see a couple of problems:

1. When all 8 connections are established, there appears to be a point where there is a lag in processing the channel requests for the 7th and 8th connections. As a result the request arriving at the ISO-AL appears to be delayed by more than an ISO interval or even more.

So for example if we consider before this PR and after for ISO Interval 10ms, BN=1, FT=1 :

**Before:**
T=12ms   Central       HCI:  SDU 1  ----- ISO-AL: Target Event 2 -----> PDU 2
T=20ms   Peripheral RX:    PDU 2  ----- ISO-AL: --------------------> SDU 2
T=22ms   Central       HCI:  SDU 2  ----- ISO-AL: Target Event 3 -----> PDU 3
T=30ms   Peripheral RX:    PDU 3  ----- ISO-AL: --------------------> SDU 3
T=32ms   Central       HCI:  Delayed
T=40ms   Peripheral RX:    FLUSH  ----- ISO-AL: --------------------> SDU 4 (No data Err) **
T=42ms   Central       HCI:  SDU 3  ----- ISO-AL: Target Event 5 -----> PDU 5
T=50ms   Peripheral RX:    PDU 5  ----- ISO-AL: --------------------> SDU 5 (Skewed, should have been SDU 4)
T=52ms   Central       HCI:  SDU 4  ----- ISO-AL: Target Event 6 -----> PDU 6
T=60ms   Peripheral RX:    PDU 6  ----- ISO-AL: --------------------> SDU 6 (Skewed, should have been SDU 5)

**After:**
T=12ms   Central       HCI:  SDU 1  ----- ISO-AL: Target Event 2 -----> PDU 2
T=20ms   Peripheral RX:    PDU 2  ----- ISO-AL: --------------------> SDU 2
T=22ms   Central       HCI:  SDU 2  ----- ISO-AL: Target Event 3 -----> PDU 3
T=30ms   Peripheral RX:    PDU 3  ----- ISO-AL: --------------------> SDU 3
T=32ms   Central       HCI:  Delayed
T=40ms   Peripheral RX:    FLUSH  ----- ISO-AL: --------------------> SDU 4 (No data Err)
T=42ms   Central       HCI:  SDU 3  ----- ISO-AL: Target Event 5 -----> PDU 3 [No skew, but too late for TX so flushed]
T=50ms   Peripheral RX:    FLUSH  ----- ISO-AL: --------------------> SDU 5 (No data Err)
T=52ms   Central       HCI:  SDU 4  ----- ISO-AL: Target Event 6 -----> PDU 4 [No skew, but too late for TX so flushed]
T=60ms   Peripheral RX:    FLUSH  ----- ISO-AL: --------------------> SDU 6 (No data Err)

2.  On the peripheral side, if only null PDUs are received, I don't see the error status of the PDUs sent to ISO-AL being set. This seems to be transformed in to a Start Continue PDU with zero length and no error status so it is interpreted as a padding PDU by the ISO-AL and no SDU is released. This means that in the previous example, the SDUs with status "No data Err" would not be received by the application.

Reply:
The "No Data Err" in the previous comment is actually "Lost Data" `ISOAL_PDU_STATUS_LOST_DATA`.